accounts/abi/bind/bind.go

```diff
@@ -0,0 +1,395 @@
+// Copyright 2016 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package bind generates Ethereum contract Go bindings.
+package bind
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"golang.org/x/tools/imports"
+)
+
+// Bind generates a Go wrapper around a contract ABI. This wrapper isn't meant
+// to be used as is in client code, but rather as an intermediate struct which
+// enforces compile time type safety and naming convention opposed to having to
+// manually maintain hard coded strings that break on runtime.
+func Bind(abijson string, bytecode string, pkg string, kind string) (string, error) {
+	// Parse the actual ABI to generate the binding for
+	abi, err := abi.JSON(strings.NewReader(abijson))
+	if err != nil {
+		return "", err
+	}
+	// Generate the contract type, fields and methods
+	code := new(bytes.Buffer)
+	kind = strings.ToUpper(kind[:1]) + kind[1:]
+	fmt.Fprintf(code, "%s\n", bindContract(kind, strings.TrimSpace(abijson), strings.TrimSpace(bytecode)))
+
+	methods := make([]string, 0, len(abi.Methods))
```

Comment:
Generating a binding? Not sure about it. Originally the entire thing was in ABI, that could have worked, but it starts introducing weird dependencies. The ABI package should really only have to do with calling conventions, code generation is a bit pushing it. I think having `bind` as a subpackage should be enough to make it clear that it's *part* of the ABI, but not an essential thing for usual use cases.